subgraphs/_Metrics/BlockSpeedForecast.ts

```diff
@@ -0,0 +1,142 @@
+////////////////////////
+// NAME: CHRIS STEEGE //
+// VERSION 1.0.0      //
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// The purpose of this program is to dynamically estimate the blocks generated for the 24 HR period following the most recent update. //  
+// It does so by calculating the moving average block rate for an arbitrary length of time preceding the current block.               //
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// Schema entities necessary for getting moving average block speed
+
+// type _BlockTracker @entity {
+//     id: ID! 
+//     blocks: [_Block!]!
+//     blocksPerDay: BigDecimal!
+// }
+
+// type _Block @entity {
+//   id: ID!
+//   blockNumber: Int!
+//   timestamp: Int!
+// }
+
+import { BigDecimal, ethereum, store } from "@graphprotocol/graph-ts";
+import { _Block, _BlockTracker } from "../../generated/schema";
+import { BIGDECIMAL_ZERO, INT_ONE, INT_TWO, INT_ZERO } from "./constants";
+
+// Forecast period. This gives you the time period that you want to estimate count of blocks per interval, based on moving average block speed.
+// 86400 = 1 Day
+export const RATE_IN_SECONDS = 86400
+export const RATE_IN_SECONDS_BD = BigDecimal.fromString(RATE_IN_SECONDS.toString())
+
+// WINDOW_SIZE_SECONDS and BLOCK_STORAGE_INTERVALS can be modified. These are just recommended values - 'somewhat' arbitrary. 
+
+// The window size determines the range of blocks that you track from the current block minus the window size.
+// Window of block time used to calculate the moving average.
+// Increasing means less deviation but also less sensitivity to changing block speeds.
+export const WINDOW_SIZE_SECONDS = 86400
+export const WINDOW_SIZE_SECONDS_BD = BigDecimal.fromString(WINDOW_SIZE_SECONDS.toString())
+
+// The storage interval tells you to only store blocks that are space by at least this amount. 
+// Increasing this value will mean less blocks stored and less frequently computes blocksPerDay.
+export const BLOCK_STORAGE_INTERVAL = 5 as i32
+
+// Paramaters: [event.block.timestamp, event.block.number]
+// Call this function in event handlers frequently enough so that it updates on most blocks 
+export function updateMovingAverageBlockSpeedForecast(event: ethereum): void {
+
+    let blockTracker = getOrCreateBlockTracker()
+    let blocks: string[]
+    blocks = blockTracker.blocks
+
+    // Create entity for the current block
+    let currentTimestamp = event.block.timestamp.toI32()
+    let currentBlockNumber = event.block.number.toI32()
+    let currentBlock = getOrCreateBlock(currentTimestamp, currentBlockNumber)
+
+    if (blocks.length == INT_ZERO) {
+        blocks.push(currentBlock.id)
+        blockTracker.blocks = blocks
```

Comment:
don't you need to update `blocksPerDay` here? Isn't it still `BIG_DECIMAL_ZERO`?

Reply:
The only time the length should be zero is for the first block. Which in this case you don't have an. interval to compute the rate.

Reply:
I think it's more accurate to use a default here (e.g. chain average) here instead of zero